### PR TITLE
martian: fix half-close propagation downstream when using http2 handler

### PR DESCRIPTION
**Pull request by mmatczuk**

### CODE REVIEW

File: internal/martian/errors.go [hunk 1]
```diff
@@ -45,7 +45,8 @@ func isClosedConnError(err error) bool {
 	if errors.Is(err, io.EOF) ||
 		errors.Is(err, io.ErrUnexpectedEOF) ||
 		errors.Is(err, syscall.ECONNABORTED) ||
-		errors.Is(err, syscall.ECONNRESET) {
+		errors.Is(err, syscall.ECONNRESET) ||
+		errors.Is(err, errClosedBody) {
```
Comment (Choraden) [inline internal/martian/errors.go:49]: Yes! I've seen a lot of these recently.

File: internal/martian/unsafe.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2022-2024 Sauce Labs Inc., all rights reserved.
+
+package martian
+
+import (
+	_ "unsafe"
+)
+
+//go:linkname errClosedBody golang.org/x/net/http2.errClosedBody
+var errClosedBody error
```
Comment (Choraden) [inline internal/martian/unsafe.go:10]: Isn't that file name too general?

File: internal/martian/proxy_handler.go [hunk 3]
```diff
@@ -337,27 +337,36 @@ func (p proxyHandler) handleRequest(rw http.ResponseWriter, req *http.Request) {
...
-	rc *http.ResponseController
+type h2Writer struct {
+	w     io.Writer
+	flush func() error
+	close func() error
 }
 
-func (w writeFlusher) Write(p []byte) (n int, err error) {
-	n, err = w.rw.Write(p)
+func makeH2Writer(rw http.ResponseWriter, rc *http.ResponseController, req *http.Request) *h2Writer {
```
Comment (Choraden) [inline internal/martian/proxy_handler.go:346]: Return a pointer? Previous flusher was a copy, and the methods receive a copy too, so I'm double checking if that's on purpose.

File: internal/martian/unsafe.go (hunk 2 above)
Comment (mmatczuk) [reply internal/martian/unsafe.go:10]: Initially I called it h2Err... but reverted to that due to linter

File: internal/martian/proxy_handler.go (hunk 3 above)
Comment (mmatczuk) [reply internal/martian/proxy_handler.go:346]: Nah, that's too fast generation...

File: internal/martian/unsafe.go (hunk 2 above)
Comment (Choraden) [reply internal/martian/unsafe.go:10]: Yeah, I was thinking of some `h2error.go`

File: .golangci.yml [hunk 4]
```diff
@@ -141,6 +141,10 @@ issues:
         - golint
       source: 'import _ "embed"'
 
+    - linters:
+        - revive
+      source: '_ "unsafe"'
```
Comment (Choraden) [inline .golangci.yml:146]: grep for already existing comments

File: internal/martian/proxy_handler.go (hunk 3 above)
Comment (mmatczuk) [reply internal/martian/proxy_handler.go]: Fixed

File: internal/martian/unsafe.go (hunk 2 above)
Comment (mmatczuk) [reply internal/martian/unsafe.go]: go:linkname moved to errors.go and renamed to `h2ErrClosedBody`.

File: .golangci.yml (hunk 4 above)
Comment (mmatczuk) [reply .golangci.yml]: Dropped that commit.

**Review by Choraden — APPROVED** [review]
LGTM

**Pull request merged**